src/Sia.Gateway/Initialization/Startup/ApplicationInsightsStartup.cs

```diff
@@ -0,0 +1,37 @@
+﻿using Microsoft.AspNetCore.Hosting;
+using Microsoft.Extensions.Configuration;
+using Sia.Gateway.Initialization.Configuration;
+using Sia.Shared.Authentication;
+using Sia.Shared.Validation;
+
+namespace Sia.Gateway.Initialization
+{
+    public static class ApplicationInsightsStartup
+    {
+        public static void InitializeApplicationInsights(this IHostingEnvironment env, IConfigurationRoot configuration)
+        {
+            //Needs to be done in the initial Startup.Startup() method because Application Insights registers itself prior
+            //to ConfigureServices being run
+            var deserializedConfig = configuration.Get<GatewayConfiguration>();
+
+            var secrets = new AzureSecretVault(
+                new KeyVaultConfiguration(
+                    deserializedConfig.ClientId,
+                    deserializedConfig.ClientSecret,
+                    deserializedConfig.KeyVault.VaultName
+                )
+            );
+
+            var instrumentationKeyName = ThrowIf.NullOrWhiteSpace(deserializedConfig.KeyVault.InstrumentationKeyName, nameof(deserializedConfig.KeyVault.InstrumentationKeyName));
+
+            var vaultTask = secrets.Get(instrumentationKeyName);
+            vaultTask.Wait();
```

Comment:
Ideally, you don't want to block the thread here. Instead, you push the waiting to the most outer layer, where async/await is not feasible.

Additionally, since thread context switching is not expected to break stuff here (e.g. there is no UI thread to sync the context with), you should add ConfigureAwait(continueOnCapturedContext: false) to avoid possible deadlocks. (Note that context synchronization is not an issue anymore with ASP.NET Core, but it's good practice to disable continueOnCapturedContext unless it's needed, in case this method gets called by a library and cause a deadlock).

`var instrumentationKey = await secrets.Get(instrumentationKeyName).ConfigureAwait(continueOnCapturedContext: false);`

Reply:
Made some changes; pushed task tracking out into the Startup method and made InitializeApplicationInsights async. Let me know if there's any other changes needed for async on this.